picoCTF-web/api/team.py

```diff
@@ -190,6 +190,10 @@ def create_team(params):
     params['size'] = 0
     params['instances'] = {}
 
+    settings = api.config.get_settings()
+    if settings["shell_servers"]["enable_sharding"]:
+        params['server_number'] = api.shell_servers.assign_server_number()
+
```

Comment:
Currently trying to understand the code. (So I may not fully understand the changes.)

Wouldn't it be better to assign `params["server_number"]` to `1` when sharding is disabled instead of not having this key in `params`? This way, the docs in the teams collection will all have the key `server_number` from the very beginning. So when the platform admin decides to enable sharding *at a later date*, the admin can just set a step at the current number of teams and restart the server. The existing teams will all go shard 1, and all future teams will go shard 2?

(It seems the current code would entail the admin needing to go into mongo to insert the key `server_number` for all existing teams in the "enable sharding at a later date" scenario.)

Reply:
That's fine too, default `server_num` 1. My thought process went towards the (admittedly) minor factor of keeping team doc cleaner if sharding was disabled. Then something like 
`team.get('server_number', 1)` when needed.

Reply:
But this concept of `team.get('server_number', 1)` is not implemented in this PR, right?

Reply:
Added safe get()'s in 1b4ad4a

I'm considering scenarios in which one may be toggling sharding on and off in between team creations, adding servers, etc and the `server_number` key is omitted. 

I think it's better that setters not default to 1 (with exception of server1 when server count is 0) to be able to more clearly identify entries added during time that sharding was disabled, so some mongodb-level cleanup can be done by admin. 

Getters on otherhand, can default to 1 when keys don't exist.

Reply:
I agree that supporting initial-off-then-on is the best approach.

Initial sharding off:
* team` server_number` is omitted on `create_team()`. I believe this is better to allow for targeted reassignment later. GET is irrelevant
* The first added shell_server (which may possibly be added automatically by ansible playbook, is designated as `server_number = 1` . 
* In just added ffe28edd , problem instance `server_number` will inherit from whatever was set during `add_server()`. This will likely be `1` given the previous point

Turning sharding on:

* Unassigned teams will default on `server_number` GET to `1` .  Per 311a89a, existing teams without assignment can be assigned on to `server_numbers` TODO: find where to best do this check and assignment
* First shell_server needs no change. Additional shell_servers have `server_number` input box on management interface. If a second server was added before sharding was enabled, and has an unspecified  `server_number`, **the server must be updated/readded and Load Deployment**. 
* Following the point above, no problem instances should be undesignated. By requiring a Load Deployment, this will insure that any shell_servers *without* a clear `server_number` designation do not get assigned to teams.

Reply:
There's also another possible other team `server_server` assignments besides the above on `create_team()`

That is, to re-assign teams (regardless of they have been already assigned) based on possibly  *new* shard stepping. This would need to be manually called.